dali/operators/segmentation/biased_crop_center.cc

```diff
@@ -0,0 +1,180 @@
+
+The purpose of this operator is to obtain a distribution of cropping window centers that
+has a certain bias towards selecting a foreground pixel as a center.
+)")
+    .AddOptionalArg("foreground",
+      R"code(If different than 0, the crop center is selected to match any of the foreground
+ pixels in the input mask. If 0, the crop center is selected randomly.)code",
+      0, true)
+    .AddOptionalArg<std::vector<int>>("shape",
+      R"code(If specified, it represents the shape of the cropping window to be used,
+introducing restrictions to the range of valid crop centers.
+
+When foreground == 0, a random cropping center is selected so that the cropping window
+is fully contained in the input.
+
+When foreground != 0, the cropping center is first picked to match any of the foreground
+pixels in the input. If the selected crop center results in an out of bounds cropping window,
+the center is shifted as necessary so that the window remains within bounds.
+
+If left unspecified, the shape is not taken into account when selecting the center,
+effectively resulting in any position in the input being a valid center.
+)code",
+      nullptr, true)
+    .NumInput(1)
+    .NumOutput(1);
+
+class BiasedCropCenterCPU : public Operator<CPUBackend> {
+ public:
+  explicit BiasedCropCenterCPU(const OpSpec &spec);
+  bool CanInferOutputs() const override { return true; }
+  bool SetupImpl(std::vector<OutputDesc> &output_desc, const workspace_t<CPUBackend> &ws) override;
+  void RunImpl(workspace_t<CPUBackend> &ws) override;
+
+ private:
+  template <typename T>
+  void RunImplTyped(workspace_t<CPUBackend> &ws);
+
+  std::mt19937_64 rng_;
+
+  std::vector<int> foreground_;
+  TensorListShape<> crop_shape_;
+
+  bool has_crop_shape_ = false;
+
+  USE_OPERATOR_MEMBERS();
+};
+
+BiasedCropCenterCPU::BiasedCropCenterCPU(const OpSpec &spec)
+    : Operator<CPUBackend>(spec), rng_(spec.GetArgument<int64_t>("seed")) {
+}
+
+bool BiasedCropCenterCPU::SetupImpl(std::vector<OutputDesc> &output_desc,
+                                    const workspace_t<CPUBackend> &ws) {
+  const auto &in_masks = ws.template InputRef<CPUBackend>(0);
+  int nsamples = in_masks.size();
+  auto in_masks_shape = in_masks.shape();
+  int ndim = in_masks_shape.sample_dim();
+  output_desc.resize(1);
+  output_desc[0].shape = uniform_list_shape(nsamples, {ndim});
+  output_desc[0].type = TypeTable::GetTypeInfo(DALI_INT64);
+
+  foreground_.clear();
+  if (spec_.HasTensorArgument("foreground")) {
+    foreground_.resize(nsamples);
+    const auto &foreground_arg_in = ws.ArgumentInput("foreground");
+    for (int i = 0; i < nsamples; i++) {
+      foreground_[i] = *foreground_arg_in[i].data<int>();
+    }
+  } else {
+    foreground_.push_back(spec_.GetArgument<int>("foreground"));
+  }
+
+  has_crop_shape_ = spec_.ArgumentDefined("shape");
+  if (has_crop_shape_) {
+    GetShapeArgument(crop_shape_, spec_, "shape", ws, ndim, nsamples);
```

Comment:
Shouldn't you do anything with the return value as well?

Reply:
not really, I don't need to infer the batch size or the number of dimensions.